Fix legend symbol size when using map units (fix #13078)

vmora opened the pull request:
The computation of icon sizes for legend symbols has been moved to QgsLayerTreeModel::legendInvalidateMapBasedData() susch that icon size is recomputed when zooming.

@wonder-sk thanks a lot for the pointers, does that look good to you ?

vmora committed:
Fix legend symbol size when using map units
fix #13078
The computation of icon sizes for legend symbols has been moved to
QgsLayerTreeModel::legendInvalidateMapBasedData() susch that icon size
is recomputed when zooming.

wonder-sk commented on src/core/layertree/qgslayertreemodel.cpp:
```diff
@@ -1299,10 +1303,41 @@ QList<QgsLayerTreeModelLegendNode*> QgsLayerTreeModel::layerLegendNodes( QgsLaye
+      QgsSymbolV2LegendNode* n = dynamic_cast<QgsSymbolV2LegendNode*>( legendNode );
+      if ( n )
+      {
+        n->setParent( this ); // map scale are in the model, so the parent needs to be set
```
is it necessary to call setParent() ? That should have been done in addLegendToLayer() already...

vmora replied:
No it's not necessary, you're right. 

I added this because when the symbol size computation was done in QgsDefaultVectorLayerLegend::createLayerTreeModelLegendNodes the parent was not set... and I wanted to make sure :)

wonder-sk commented on src/core/layertree/qgslayertreemodel.cpp:
```diff
@@ -1299,10 +1303,41 @@ QList<QgsLayerTreeModelLegendNode*> QgsLayerTreeModel::layerLegendNodes( QgsLaye
+  // the symbol sizes changes depends on the zoom level
+
+  QList<QgsSymbolV2LegendNode*> symbolNodes;
+  QMap<QString, int> widthMax;
```
widthMax should work per-layer - otherwise it may overwrite keys among different layers (especially root key which is empty string)

vmora replied:
I moved the 2 lines inside the loop over layers, and moved the 
  
  foreach ( QgsSymbolV2LegendNode* n, symbolNodes )

inside as well.

wonder-sk commented on src/core/layertree/qgslayertreemodel.cpp:
```diff
@@ -1299,10 +1303,41 @@ QList<QgsLayerTreeModelLegendNode*> QgsLayerTreeModel::layerLegendNodes( QgsLaye
+    Q_ASSERT( widthMax[parentKey] > 0 );
+    const int twiceMarginWidth = 2; // a one pixel margin avoids hugly rendering of icon
+    n->setIconSize( QSize( widthMax[parentKey] + twiceMarginWidth, n->iconSize().rheight() + twiceMarginWidth ) );
+    n->invalidateMapBasedData();
```
is it necessary to call invalidateMapBasedData() again here?

vmora replied:
I don't think so, I'll test is out.

vmora replied:
It is apparently, at least it seems necessary to call that last (after set iconSize).

vmora committed:
fixups (to be rebased)
- removed the setParent since the parent is already set here
- moved the legendNode->invalidateMapBasedData() after size setting
  (removed dupplicate call)
- moved max icon size computation in the loop over layers

vmora commented:
@wonder-sk is it ok for you now ?

wonder-sk commented:
Looks good - thanks!

SebDieBln commented:
@vmora @wonder-sk I would like to draw your attention to [Redmine #14172](https://hub.qgis.org/issues/14172) about a regression introduced by this PR.

Is there any way to determine in advance whether a symbol actually uses map based data?

At the moment all symbols seem to be recalculated after every pan or zoom and QGIS freezes during this time. It is not noticable when dealing with less than hundred symbols or so, but with more than thousand symbols (due to some layers with graduated styles) it adds up to a second or more, even on decent hardware.
